core/src/main/java/com/digitalpebble/storm/crawler/bolt/FetcherBolt.java

```diff
@@ -336,42 +324,61 @@ public synchronized FetchItemQueue getFetchItemQueue(String id) {
                 fiq = new FetchItemQueue(conf, customThreadVal, crawlDelay,
                         minCrawlDelay);
                 queues.put(id, fiq);
-
-                // Reset the cyclic iterator to start of the list.
-                it = Iterables.cycle(queues.keySet()).iterator();
             }
             return fiq;
         }
 
         public synchronized FetchItem getFetchItem() {
-
-            if (queues.isEmpty() || !it.hasNext())
+            if (queues.isEmpty()) {
                 return null;
+            }
 
             FetchItemQueue start = null;
 
             do {
-                FetchItemQueue fiq = queues.get(it.next());
+                Map.Entry<String, FetchItemQueue> nextEntry;
+                FetchItemQueue fiq;
+                Iterator i = queues.entrySet().iterator();
```

Comment:
How costly is this?

You need to synchronize on the map when iterating over the entry set - [https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedMap(java.util.Map)]